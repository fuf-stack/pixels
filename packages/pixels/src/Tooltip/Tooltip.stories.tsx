import type { Meta, StoryObj } from '@storybook/react-vite';
import type { TooltipPlacement, TooltipProps } from './Tooltip';

import Tooltip, { tooltipVariants as heroTooltipVariants } from './Tooltip';

const meta: Meta<typeof Tooltip> = {
  title: 'pixels/Tooltip',
  component: Tooltip,
};

export default meta;
type Story = StoryObj<typeof Tooltip>;

const sizes = Object.keys(heroTooltipVariants.variants.size);

export const Default: Story = {
  args: {
    content: 'tooltip content',
    children: 'hover me',
  },
};

export const Delay: Story = {
  args: {
    content: 'I took 1 second',
    children: '1000 ms delay',
    delay: 1000,
  },
};

export const CloseDelay: Story = {
  args: {
    content: 'I will go in 3s',
    children: '3000 ms close delay',
    closeDelay: 3000,
  },
};

const tooltipPlacementOptions: TooltipPlacement[] = [
  'top',
  'bottom',
  'left',
  'right',
  'top-start',
  'top-end',
  'bottom-start',
  'bottom-end',
  'left-start',
  'left-end',
  'right-start',
  'right-end',
];

export const AllPlacements: Story = {
  render: (args) => (
    <>
      {tooltipPlacementOptions.map((placement) => (
        <div key={placement} className="mb-6">
          <Tooltip placement={placement} {...args}>
            {placement}
          </Tooltip>
        </div>
      ))}
    </>
  ),
  args: {
    content: 'tooltip content',
  },
};

export const TooltipInTooltip: Story = {
  args: {
    content: (
      <Tooltip content="2nd tooltip content">
        hover for 2nd tooltip content
      </Tooltip>
    ),
    children: 'hover me',
  },
};

export const Sizes: Story = {
  args: {
    content:
      'tooltip content. this can be very long. Or not. Depending on what you want. Just make sure to test it! To test it, we need a longer text. So I will just keep on writing more words. adjuifgoreiaidofgreougdnfigerigahrisuiasoghureiidfgjjuirghuidsghuegijuifghreugdjagvioeruhgidugdfsihguirghreuigjaiuhgfuidofhgidhgsuihagiuohegiofdpsjguoeriirgidosfhguihoeg',
    children: 'hover me',
  },
  render: (args) => {
    return (
      <div className="flex flex-col gap-4">
        {sizes.map((size) => (
          <Tooltip key={size} size={size as TooltipProps['size']} {...args}>
            {size}
          </Tooltip>
        ))}
      </div>
    );
  },
};

export const DefaultOpen: Story = {
  args: {
    content: 'tooltip content',
    footer: 'footer',
    header: 'header',
    children: 'hover me',
    defaultOpen: true,
  },
};
