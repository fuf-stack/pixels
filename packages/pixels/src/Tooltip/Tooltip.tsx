import type { TVClassName, TVProps } from '@fuf-stack/pixel-utils';
import type { TooltipPlacement as HeroTooltipPlacement } from '@heroui/tooltip';
import type { ReactNode } from 'react';

import { Tooltip as HeroTooltip } from '@heroui/tooltip';

import { tv, variantsToClassNames } from '@fuf-stack/pixel-utils';

// tooltip variants
export const tooltipVariants = tv({
  slots: {
    base: '',
    content: '',
    wrapper: 'cursor-pointer',
  },
});

type VariantProps = TVProps<typeof tooltipVariants>;
type ClassName = TVClassName<typeof tooltipVariants>;

export type TooltipPlacement = HeroTooltipPlacement;

export interface TooltipProps extends VariantProps {
  /** trigger child components */
  children: ReactNode;
  /** CSS class name */
  className?: ClassName;
  /** delay in milliseconds before the tooltip opens. */
  closeDelay?: number;
  /** delay in milliseconds before the tooltip opens. */
  delay?: number;
  /** placement padding in px */
  containerPadding?: number;
  /** content displayed in the tooltip */
  content: ReactNode;
  /** open overlay initially when uncontrolled */
  defaultOpen?: boolean;
  /** handler that is called when the overlay's open state changes */
  onOpenChange?: (isOpen: boolean) => void;
  /** placement if the tooltip */
  placement?: TooltipPlacement;
}

/**
 * Tooltip component based on [HeroUI Tooltip](https://www.heroui.com//docs/components/tooltip)
 */
const Tooltip = ({
  children,
  className: _className = undefined,
  closeDelay = 500,
  containerPadding = 0,
  content,
  defaultOpen = false,
  delay = 0,
  onOpenChange = undefined,
  placement = 'top',
}: TooltipProps) => {
  // classNames from slots
  const variants = tooltipVariants();
  const classNames = variantsToClassNames(variants, _className, 'base');

  return (
    <HeroTooltip
      classNames={classNames}
      closeDelay={closeDelay}
      containerPadding={containerPadding}
      content={content}
      defaultOpen={defaultOpen}
      delay={delay}
      onClick={(e) => e.preventDefault()}
      onOpenChange={onOpenChange}
      placement={placement}
      shouldFlip
      showArrow
    >
      <span className={classNames.wrapper}>{children}</span>
    </HeroTooltip>
  );
};

export default Tooltip;
