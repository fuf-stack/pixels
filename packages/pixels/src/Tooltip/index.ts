import Tooltip, { tooltipPlacementOptions } from './Tooltip';

export type { TooltipProps } from './Tooltip';

export { Tooltip, tooltipPlacementOptions };

export default Tooltip;
