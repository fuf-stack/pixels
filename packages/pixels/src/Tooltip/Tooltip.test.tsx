import { describe } from 'vitest';

import storySnapshots from 'storybook-config/story-snapshots';

import * as stories from './Tooltip.stories';

describe('Story Snapshots', () => {
  storySnapshots(stories);
});
