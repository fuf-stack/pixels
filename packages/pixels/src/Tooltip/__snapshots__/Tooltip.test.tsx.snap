// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Story Snapshots > AllPlacements 1`] = `
<div>
  <div
    class="mb-6"
  >
    <span
      class="cursor-pointer"
    >
      top
    </span>
  </div>
  <div
    class="mb-6"
  >
    <span
      class="cursor-pointer"
    >
      bottom
    </span>
  </div>
  <div
    class="mb-6"
  >
    <span
      class="cursor-pointer"
    >
      left
    </span>
  </div>
  <div
    class="mb-6"
  >
    <span
      class="cursor-pointer"
    >
      right
    </span>
  </div>
  <div
    class="mb-6"
  >
    <span
      class="cursor-pointer"
    >
      top-start
    </span>
  </div>
  <div
    class="mb-6"
  >
    <span
      class="cursor-pointer"
    >
      top-end
    </span>
  </div>
  <div
    class="mb-6"
  >
    <span
      class="cursor-pointer"
    >
      bottom-start
    </span>
  </div>
  <div
    class="mb-6"
  >
    <span
      class="cursor-pointer"
    >
      bottom-end
    </span>
  </div>
  <div
    class="mb-6"
  >
    <span
      class="cursor-pointer"
    >
      left-start
    </span>
  </div>
  <div
    class="mb-6"
  >
    <span
      class="cursor-pointer"
    >
      left-end
    </span>
  </div>
  <div
    class="mb-6"
  >
    <span
      class="cursor-pointer"
    >
      right-start
    </span>
  </div>
  <div
    class="mb-6"
  >
    <span
      class="cursor-pointer"
    >
      right-end
    </span>
  </div>
</div>
`;

exports[`Story Snapshots > CloseDelay 1`] = `
<div>
  <span
    class="cursor-pointer"
  >
    3000 ms close delay
  </span>
</div>
`;

exports[`Story Snapshots > Default 1`] = `
<div>
  <span
    class="cursor-pointer"
  >
    hover me
  </span>
</div>
`;

exports[`Story Snapshots > Delay 1`] = `
<div>
  <span
    class="cursor-pointer"
  >
    1000 ms delay
  </span>
</div>
`;
